`SideNav` - Add missing "Component API" docs and warning about refocusing behaviour (#1551)
hashicorp/design-system: website/docs/components/side-nav/partials/code/component-api.md

  <C.Property @name="<:footer>" @type="named block">
    A named block where the content for the “footer” section of the SideNav is rendered. It yields the value of `isMinimized` too.
  </C.Property>

>   <C.Property @name="isResponsive" @type="boolean" @default="true">

Reviewer: I might be mixing it up with the previous side nav, so let's just double check that the outcomes are what we want: 

1. when zoomed into 400% it reflows properly
2. scrolling up/down is possible and sticky content doesn't hide other interactive elements

The reason that I mention a zoom percent instead of pixels is the same reason the spec explainer doc has this note: 
![CleanShot 2023-08-08 at 14 33 07](https://github.com/hashicorp/design-system/assets/4587451/88c35670-3721-4884-adca-a2a647cdb6f9)

>   <C.Property @name="isResponsive" @type="boolean" @default="true">

Author: > 1. when zoomed into 400% it reflows properly

See screenshot attached above.

> 2. scrolling up/down is possible 

It is already possible. But in this case, is there a specific height to support? (256px is frankly impossible to satisfy)

> and sticky content doesn't hide other interactive elements

With "other interactive elements" do you mean inside the sidenav (in which case it may happen for certain very small heights) or in the main content of the page (in which case at mobile viewports it's inevitable, since it works as overlay, triggered by the user)

>   <C.Property @name="isResponsive" @type="boolean" @default="true">

Reviewer: Inside the sidenav, yes. But I will file a separate issue to address this, because not only does it not have to be inevitable, but to leave it that way is to knowingly ship something that isn't WCAG conformant.

    Controls whether the SideNav is responsive to viewport changes. It can be programmatically turned off passing `false`.
    <br>

>     <em>Notice: even if the @isResponsive parameter is set to false, some JavaScript is executed anyway in the background, and events listeners are attached to some DOM elements (even if these functionality are not used).</em>

Reviewer 2: ```suggestion
    <em>Notice: even if the @isResponsive parameter is set to false, some JavaScript is executed anyway in the background, and event listeners are attached to some DOM elements (even if this functionality is not used).</em>
```

  </C.Property>

>   <C.Property @name="hasA11yRefocus" @type="boolean" @default="true">

Reviewer 3: [question] do we know the ramifications for accessibility if people opt out of this (which I suspect could be common if it's the most effective way to work around the issues folks were reporting). My fear would be that once people opt out of this, getting them to opt back in could be difficult (or impossible).

Author: Good point. Probably we should document it. @MelSumner can you add a paragraph or two about this in the PR?

Reviewer 3: [🌶️  question] _should_ we let consumers opt out (in an ideal world)?

Author: They may decide to implement their own solution. Or they may be already using another one. Or they may have good reasons for not using it. In any case, it's always easier to have a simple switch, than having them hack something around. _Guidance over control_.

Reviewer 4: Arguably it is easier to see who has this (or any feature for that matter) turned off and potentially work with them to find out why and how we can support them to align with our recommendations, than not providing it and being unable to know who has workarounds.

Reviewer 4: That being said, introducing this argument means that the component may become conditionally conformant – so we should probably flag that in the Accessibility section

Reviewer: if `isResponsive` is set to true, then the component will not pass the Reflow Success Criteria.

Reviewer: > [🌶️ question] _should_ we let consumers opt out (in an ideal world)?

Alas we are not in an ideal world. We have already set a precedent for a non-conformant escape hatch (see FlightIcon docs) and here we are continuing that precedent. However, we could use stronger language like we did before, that indicates that these API are more like "if you must, while you're transitioning to HDS components and working through your a11y issues at an application level."

>     Controls whether a "navigator narrator" and a "skip link" are added to the navigation (provided by the [`ember-a11y-refocus` Ember addon](https://github.com/ember-a11y/ember-a11y-refocus)). It can be programmatically turned off passing `false`.

Reviewer 2: ```suggestion
    Controls whether a "navigator narrator" and a "skip link" are added to the navigation (provided by the [`ember-a11y-refocus` Ember addon](https://github.com/ember-a11y/ember-a11y-refocus)). It can be programmatically turned off by passing `false`.
```

>     Controls whether a "navigator narrator" and a "skip link" are added to the navigation (provided by the [`ember-a11y-refocus` Ember addon](https://github.com/ember-a11y/ember-a11y-refocus)). It can be programmatically turned off by passing `false`.

Reviewer: ```suggestion
    Controls whether a "navigator narrator" and a "skip link" are added to the navigation (provided by the [`ember-a11y-refocus` Ember addon](https://github.com/ember-a11y/ember-a11y-refocus)). It can be programmatically turned off by passing `false`. Warning: if it is set to false, then it will fail Bypass Blocks, [Success Criteria 2.4.1](https://www.w3.org/WAI/WCAG21/Understanding/bypass-blocks.html). Since this component appears on every page, the application will not be considered conformant.
```

    <br><br>
    <em>For details about the addon behaviour and functionality, refer to the [official documentation](https://github.com/ember-a11y/ember-a11y-refocus#readme).</em>
    <Doc::ComponentApi as |C|>
      <C.Property @name="a11yRefocusSkipTo" @type="string">
        Pass through property for the `skipTo` argument - The element ID that should receive focus on skip.
      </C.Property>
      <C.Property @name="a11yRefocusSkipText" @type="string">

>         Pass through property for the `skipText` argument - The text passed in the skip link; defaults to "Skip to main content".

Reviewer 2: ```suggestion
        Pass-through property for the `skipText` argument—The text passed in the skip link; defaults to "Skip to main content".
```

      </C.Property>
      <C.Property @name="a11yRefocusNavigationText" @type="string">

>         Pass through property for the `navigationText` argument - The text used as navigation message. Defaults to "The page navigation is complete. You may now navigate the page content as you wish.".

Reviewer 2: ```suggestion
        Pass-through property for the `navigationText` argument—The text used as navigation message. Defaults to "The page navigation is complete. You may now navigate the page content as you wish.".
```

      </C.Property>
      <C.Property @name="a11yRefocusRouteChangeValidator" @type="string">

>         Pass through property for the `routeChangeValidator` argument - Custom function used to define which route changes should trigger the refocusing behavior for the navigator narrator. - For details see [Customizing the definition of a route change](https://github.com/ember-a11y/ember-a11y-refocus#customizing-the-definition-of-a-route-change).

Reviewer 2: ```suggestion
        Pass-through property for the `routeChangeValidator` argument—Custom function used to define which route changes should trigger the refocusing behavior for the navigator narrator.—For details see [Customizing the definition of a route change](https://github.com/ember-a11y/ember-a11y-refocus#customizing-the-definition-of-a-route-change).
```
